packages/components/src/components/cells/code.tsx

```diff
@@ -0,0 +1,636 @@
+
+import { useCallback, useEffect, useRef, useState } from 'react';
+import { Dialog, DialogContent } from '../ui/dialog';
+import { Tooltip, TooltipContent, TooltipProvider, TooltipTrigger } from '../ui/tooltip';
+import Shortcut from '../keyboard-shortcut';
+import { useNavigate } from 'react-router-dom';
+import CodeMirror from '@uiw/react-codemirror';
+import { javascript } from '@codemirror/lang-javascript';
+import { ResizableHandle, ResizablePanel, ResizablePanelGroup } from '../ui/resizable';
+import {
+  Info,
+  Play,
+  Trash2,
+  Sparkles,
+  X,
+  MessageCircleWarning,
+  LoaderCircle,
+  Maximize,
+  Minimize,
+  CopyIcon,
+} from 'lucide-react';
+import TextareaAutosize from 'react-textarea-autosize';
+import AiGenerateTipsDialog from '../ai-generate-tips-dialog';
+import { CellType, CodeCellType, CodeCellUpdateAttrsType } from '@srcbook/shared';
+import { cn } from '../../lib/utils';
+import { CellModeType, SessionType } from '../../types';
+import { Button } from '../ui/button';
+import { Input } from '../ui/input';
+import DeleteCellWithConfirmation from '../delete-cell-dialog';
+import { CellOutput } from '../cell-output';
+import useTheme from '../use-theme';
+import { useDebouncedCallback } from 'use-debounce';
+import { EditorView } from 'codemirror';
+import { EditorState, Extension } from '@codemirror/state';
+import { unifiedMergeView } from '@codemirror/merge';
+import { toast } from 'sonner';
+import { PrettierLogo } from '../logos';
+import { useCells } from '../use-cell';
+
+const DEBOUNCE_DELAY = 500;
+
+type BaseProps = {
+  session: SessionType;
+  cell: CodeCellType;
+  codeTheme: Extension;
+};
+
+type RegularProps = BaseProps & {
+  readOnly?: false;
+  aiEnabled: boolean;
+  aiFixDiagnostics: (diagnostics: string) => void;
+  cellMode: CellModeType;
+  // channel: SessionChannel;
+  filenameError: string | null;
+  fixDiagnostics: (diagnostics: string) => void;
+  fullscreen: boolean;
+  generationType: 'edit' | 'fix';
+  newSource: string;
+  onAccept: () => void;
+  onChangeCellModeType: (newCellMode: CellModeType) => void;
+  onChangeFilenameError: (newError: string | null) => void;
+  onChangeFullscreen: (value: boolean) => void;
+  onChangeGenerationType: (newGenerationType: 'edit' | 'fix') => void;
+  onChangeNewSource: (newSource: string) => void;
+  onChangePrompt: (newPrompt: string) => void;
+  onChangeShowStdio: (showStdio: boolean) => void;
+  onDeleteCell: (cell: CellType) => void;
+  onFormatCell: () => void;
+  onGenerate: () => void;
+  onGetDefinitionContents: (pos: number, cell: CodeCellType) => Promise<string>;
+  onRevert: () => void;
+  onRunCell: () => void;
+  onStopCell: () => void;
+  onUpdateFileName: (filename: string) => void;
+  prompt: string;
+  showStdio: boolean;
+  updateCellOnServer: (cell: CodeCellType, attrs: CodeCellUpdateAttrsType) => void;
+  editorExtensions: Array<Extension>;
```

Comment:
Ben and I talked about this I beleive, but a big win in removing complexity was to move all the extension setup out of the code cell and up one level up, which means that all the tendrils that the editor specific overlays / autocomplete / etc have into the tsserver can be sidestepped in this abstraction. This worked out well!

Reply:
This is good in that all arguments to this function are now made explicit. Depending on how many levels deep this component tree goes, there might be a lot of noise as a result of the prop drilling. If this becomes tedious, we might be able to:

1. Pass all props explicitly into the root
2. Have the root create a context+hook with all the props
3. Remove some prop passing and allow children to utilize the hook

If that ends up being helpful and we need to reuse this component in different contexts, there may be an even better solution which is to place a dependency on a hook defined elsewhere and at build time swap out that dependency with different implementations for the different contexts.

All that said, this is a longer term thing, we shouldn't try to do this now.